### PR TITLE
Add docs for the human-readable schema format

### PR DESCRIPTION
**Pull request by shaobo-he-aws**
*Issue #, if available:*
#86 

*Description of changes:*
This PR also restructures the schema docs.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by aaronjeline — APPROVED** [review]

File: docs/collections/_schema/human-readable-schema-grammar.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
+---
+layout: default
+title: Human-Readable schema grammar
```
Comment (khieta) [inline docs/collections/_schema/human-readable-schema-grammar.md:3]: Weird capitalization. I think you want "Human-readable schema grammar" here and elswhere

File: docs/collections/_schema/human-readable-schema-grammar.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+This topic describes the grammar specification for the human-readable schema format. For a more complete description, see [Schema format](../schema/human-readable-schema-format).
+
+The grammar applies the following the conventions. Capitalized words stand for grammar productions, and lexical tokens are given in all-caps. When productions or tokens match those in the Cedar policy grammar, we use the same names (e.g., `IDENT` and `Path`).
+
+For grammar productions it uses `|` for alternatives, `[]` for optional productions, `()` for grouping, and `{}` for repetition of a form zero or more times.
+
+Tokens are defined using regular expressions, where `[]` stands for character ranges; `|` stands for alternation; `*` , `+` , and `?` stand for zero or more, one or more, and zero or one occurrences, respectively; `~` stands for complement; and `-` stands for difference. The grammar ignores whitespace and comments.
+
+The grammar adopts the same string escaping rules as Cedar policy grammar.
```
Comment (khieta) [inline docs/collections/_schema/human-readable-schema-grammar.md:19]:
```suggestion
The grammar adopts the same string escaping rules as the Cedar policy grammar.
```

Also maybe add a link to the policy grammar.

**Review by khieta — COMMENTED** [review]
I don't know if we want to present the "human-readable" and "json" schemas as two alternate syntaxes. It seems like it would be more consistent with the presentation of policies to focus on the new schema syntax (calling it "schema syntax" rather than "human-readable schema syntax") and then have a separate page about how to encode schemas in JSON.

File: docs/collections/_schema/human-readable-schema.md [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+
+This topic describes Cedar's human-readable schema format.
+
+## Schema format {#schema-format}
+
+A schema consists of zero or more namespaces, each of which contains declarations of three types --- *Entity Type Declaration*, *Action Declaration*, and *Common Type Declaration*. These declarations define entity types, actions, and common types used to define the former two, respectively. Declarations are delimited by `;`s. Note that unlike the JSON schema format, the human-readable schema format allows you to write Cedar-style comments.
+
+## NameSpace {#schema-namespace}
+
+You can group the declarations of a namespace by wrapping them with curly braces like `namespace Foo {...}`. A name must be given for a namespace specified by such syntax. An alternative way to declare a namespace is to “inline” declarations like `entity Bar; namespace Foo {...}`, where the entity declaration of `Bar` is under the same scope of namespace `Foo`. Names of these declarations are always referred as they are whereas declaration names of a `namespace` construct must be referred as their fully-qualified forms in other namespaces.
```
Comment (john-h-kastner-aws) [inline docs/collections/_schema/human-readable-schema.md:27]: I found the wording here confusing. Maybe:

---

You can create a namespace with associated declarations using the `namespace` key word like `namespace Foo { entity Bar; }`. The name for a namespace must be an identifier as specified in Cedar's syntax. Anything declared in this namespace must be referred to in its fully-qualified form when referenced outside of the namespace, so the declared entity type would be `Foo::Bar`.

You can instead place declaration directly outside of any namespace like `entity Bar;`. Names of these declarations are always referred to without qualification.

File: docs/collections/_schema/human-readable-schema.md [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+
+## Common types {#schema-commonTypes}
+
+Like the JSON schema format, human-readable schema syntax allows for declarations of common types so that entity type declarations can use them to avoid error-prone duplication. The syntax of common type declarations is similar to defining type aliases in most programming languages: `type <Id> = <Type>` . The right hand side of `=` is a schema type name except for common types to avoid definitional circularity.
+
+## Type name disambiguation
+
+Type names in the human-readable schema format can conflict with each other. For example, `ipaddr` is a valid unqualified common type name as well as an extension type name. `Foo::Bar` is a valid qualified common type name and an entity type name. We use the following rules to disambiguate type names.
+
+1. Primitive and extension type names cannot alias by design.
```
Comment (john-h-kastner-aws) [inline docs/collections/_schema/human-readable-schema.md:112]: This rule is confusing. Is it just saying that we haven't made any extensions that have the same name as a primitive? Given the extension types are practically primitives from the users perspective, this is hardly worth mentioning.

**Review by john-h-kastner-aws — APPROVED** [review]
Agree that we should put something up quickly

File: docs/collections/_schema/human-readable-schema-grammar.md [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+nav_order: 2
+---
+<!-- markdownlint-disable-file MD040 -->
+
+# Grammar specification for the human-readable schema format {#schema-grammar}
+{: .no_toc }
+
+This topic describes the grammar specification for the human-readable schema format. For a more complete description, see [Schema format](../schema/human-readable-schema-format).
+
+The grammar applies the following the conventions. Capitalized words stand for grammar productions, and lexical tokens are given in all-caps. When productions or tokens match those in the Cedar policy grammar, we use the same names (e.g., `IDENT` and `Path`).
```
Comment (AMZ-brandon) [inline docs/collections/_schema/human-readable-schema-grammar.md:13]: "the same names" - The same as what?

Comment (AMZ-brandon) [reply docs/collections/_schema/human-readable-schema-grammar.md:13]: "grammar productions" and "lexical tokens" aren't defined here or on the schema format page. What do they mean?

File: docs/collections/_schema/human-readable-schema.md [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+
+Note that if you omit attribute declarations, then entities of this type do not have any attributes. This is equivalent to specifying an empty record (i.e., `{}`).
+
+
+### Schema types {#schema-types}
+
+Schema types can be used as RHS of an attribute or common type declaration. Cedar data types have their corresponding schema types: The corresponding type names of Cedar primitive data types [Boolean](../policies/syntax-datatypes.html#datatype-boolean), [String](../policies/syntax-datatypes.html#datatype-string), [Long](../policies/syntax-datatypes.html#datatype-string) are `Bool`, `String`, `Long`, respectively. An entity type or an extension type is specified by its name: The entity type name is either an identifier or identifiers separated by `::`. For example, both `User` and `ExampleCo::User` are both valid entity type names. An extension type name is just an identifier, specifically, `ipaddr` or `decimal` as of now. Since Cedar verison 3.1, we reserve the namespace `__cedar` and allow fully-qualified type names for primitive and extension types with this namespace. For example, `__cedar::ipaddr` uniquely identifies the `ipaddr` extension type.
```
Comment (AMZ-brandon) [inline docs/collections/_schema/human-readable-schema.md:58]: What's RHS?

**Comment by shaobo-he-aws** [comment]
> I don't know if we want to present the "human-readable" and "json" schemas as two alternate syntaxes. It seems like it would be more consistent with the presentation of policies to focus on the new schema syntax (calling it "schema syntax" rather than "human-readable schema syntax") and then have a separate page about how to encode schemas in JSON.

I thought about it. IIRC, we decided to use the name "Cedar schema syntax" after a transitional period during which we adopt the term "human-readable schema syntax".

File: docs/collections/_schema/human-readable-schema-grammar.md (hunk 1 above)
Comment (shaobo-he-aws) [reply docs/collections/_schema/human-readable-schema-grammar.md:3]: Done.

File: docs/collections/_schema/human-readable-schema.md (hunk 3 above)
Comment (AMZ-brandon) [reply docs/collections/_schema/human-readable-schema.md:27]: Thanks John. Worked up a bit more and added to revision draft.

File: docs/collections/_schema/human-readable-schema.md (hunk 4 above)
Comment (AMZ-brandon) [reply docs/collections/_schema/human-readable-schema.md:112]: I see you approved the PR. Anything need to be done here?

File: docs/collections/_schema/human-readable-schema.md (hunk 6 above)
Comment (john-h-kastner-aws) [reply docs/collections/_schema/human-readable-schema.md:58]: right-hand-side. common enough jargon, but worth changing for the docs

File: docs/collections/_schema/human-readable-schema.md (hunk 4 above)
Comment (shaobo-he-aws) [reply docs/collections/_schema/human-readable-schema.md:112]: I think what it meant is we won't allow new extension type names to alias with primitive types. I agree it's kind of obvious.

File: docs/collections/_schema/human-readable-schema-grammar.md (hunk 5 above)
Comment (shaobo-he-aws) [reply docs/collections/_schema/human-readable-schema-grammar.md:13]: The names on the left hand side of grammars. I'd prefer to keep "grammar productions" as it's a common technical term.

**Review by AMZ-brandon — APPROVED** [review]

**Comment by mwhicks1** [comment]
> I thought about it. IIRC, we decided to use the name "Cedar schema syntax" after a transitional period during which we adopt the term "human-readable schema syntax".

I agree with just calling it "Cedar schema syntax" eventually. In the meantime: I have seen others use the term "natural syntax". Originally, I was calling it "custom syntax." I find both of these better than "human-readable" but I don't mind too much either way.

**Comment by cdisselkoen** [comment]
The Cedar Rust APIs use "natural" (eg, [`Schema::from_file_natural()`](https://docs.rs/cedar-policy/latest/cedar_policy/struct.Schema.html#method.from_file_natural).  The Cedar CLI uses "human" because it already uses "human"/"json" to distinguish the two policy formats.

In the docs, I agree with @khieta that we should be consistent with the presentation of policies, where we just present the natural syntax as the obvious one, and have a separate page where we discuss the JSON format but don't give it equal billing.

**Pull request merged**